### PR TITLE
WT-5353 Verify data continuity between history and data store

### PR DESCRIPTION
**Pull request by boboeee**
Make an enhancement to the WT Utility tool when making verification to tables. This PR aims to verify the data continuity between the history store and data store, based on the transaction IDs and timestamps. 

To make the verification continuous we will need to check if the timestamps ranges between the historical version in the history store don't overlap.

### CODE REVIEW

File: src/btree/bt_vrfy.c [hunk 1]
```diff
@@ -280,12 +284,31 @@ __verify_key_hs(WT_SESSION_IMPL *session, WT_ITEM *key, WT_VSTUFF *vs)
 #else
         WT_UNUSED(vs);
 #endif
-    }
 
+        /* Perform Verification of data continuity */
```
Comment (jeremytay90) [inline src/btree/bt_vrfy.c]: full stop. Also what have we decided about dealing with gaps in timestamps? If all tombstones are going to be masked by cursors, how can we actually determine when a value _should_ and _shouldn't_ exist?

File: src/btree/bt_vrfy.c [hunk 2]
```diff
@@ -280,12 +284,31 @@ __verify_key_hs(WT_SESSION_IMPL *session, WT_ITEM *key, WT_VSTUFF *vs)
...
+        if (orig_start.timestamp < stop.timestamp) {
+            ts1_bp = __verify_timestamp_to_pretty_string(stop.timestamp);
+            ts2_bp = __verify_timestamp_to_pretty_string(orig_start.timestamp);
+            WT_ERR_MSG(session, WT_ERROR,
+              "Key %s has a history store stop timestamp %s newer than its start timestamp %s",
```
Comment (jeremytay90) [inline src/btree/bt_vrfy.c]: Perhaps something along the lines of: "Key %s has an overlap between a value starting at %s and a value ending at %s" to indicate clearly that they are two different values with bad timestamps. Perhaps even just printing the full start and stop time pairs for both would be nice, since it gives more context into the issue. They are also included when printing the `'page_addr_string'` fyi

File: src/btree/bt_vrfy.c [hunk 3]
```diff
@@ -280,12 +284,31 @@ __verify_key_hs(WT_SESSION_IMPL *session, WT_ITEM *key, WT_VSTUFF *vs)
...
+              __wt_buf_set_printable(session, hs_key->data, hs_key->size, vs->tmp1), ts1_bp,
+              ts2_bp);
+        }
+        if (orig_start.txnid < stop.txnid) {
+            WT_ERR_MSG(session, WT_ERROR, "Key %s has a history store stop transaction (%" PRIu64
+                                          ") newer than its start transaction (%" PRIu64 ")",
+              __wt_buf_set_printable(session, hs_key->data, hs_key->size, vs->tmp1), stop.txnid,
+              orig_start.txnid);
+        }
+        memcpy(&orig_start, &start, sizeof(WT_TIME_PAIR));
```
Comment (jeremytay90) [inline src/btree/bt_vrfy.c]: could just assign values directly like when you initialized?

File: src/btree/bt_vrfy.c [hunk 4]
```diff
@@ -947,6 +949,30 @@ __verify_txn_addr_cmp(WT_SESSION_IMPL *session, WT_REF *ref, uint32_t cell_num,
       gt ? "less than" : "greater than", txn2_name, txn2);
 }
 
+/*
+ * __verify_timestamp_to_pretty_string --
+ *     Convert a timestamp to a pretty string, utilises existing timestamp to string function
```
Comment (jeremytay90) [inline src/btree/bt_vrfy.c]: fullstop

File: src/btree/bt_vrfy.c [hunk 5]
```diff
@@ -881,36 +904,15 @@ static int
...
-        ts2_bp = "WT_TS_MAX";
-        break;
-    case WT_TS_NONE:
-        ts2_bp = "WT_TS_NONE";
-        break;
-    default:
-        ts2_bp = __wt_timestamp_to_string(ts2, ts_string[1]);
-        break;
-    }
+    ts1_bp = __verify_timestamp_to_pretty_string(ts1);
```
Comment (jeremytay90) [inline src/btree/bt_vrfy.c]: can remove these pointers also

File: src/btree/bt_vrfy.c [hunk 6]
```diff
@@ -232,21 +233,24 @@ __verify_row_leaf_page_hs(WT_SESSION_IMPL *session, WT_REF *ref, WT_VSTUFF *vs)
...
 static int
-__verify_key_hs(WT_SESSION_IMPL *session, WT_ITEM *key, WT_VSTUFF *vs)
+__verify_key_hs(WT_SESSION_IMPL *session, WT_ITEM *key, WT_VSTUFF *vs, WT_CELL_UNPACK *unpack)
 {
     WT_BTREE *btree;
     WT_CURSOR *hs_cursor;
     WT_DECL_ITEM(hs_key);
     WT_DECL_RET;
-    WT_TIME_PAIR start, stop;
+    WT_TIME_PAIR start, stop, orig_start;
```
[the remark this answers is not in the document]
Comment (jeremytay90) [reply src/btree/bt_vrfy.c]: maybe `prev` makes more sense, since this is updated as you go

File: src/btree/bt_vrfy.c (hunk 2 above)
Comment (boboeee) [reply src/btree/bt_vrfy.c]: I definitely see your point about having a more clear value and string of timestamps in the string, but I would also like to conform to original style in **__verify_page_cell**, so I'm thinking we do a compromise of both, and do a string on the lines of "In the Btree %d, page reference at %s, Key %s has a overlap of timestamp ranges between history store stop timestamp %s being newer than a more recent timestamp range having start timestamp %s", let me know what you think of that message

File: src/btree/bt_vrfy.c (hunk 5 above)
Comment (boboeee) [reply src/btree/bt_vrfy.c]: agreed.

File: src/btree/bt_vrfy.c (hunk 4 above)
Comment (boboeee) [reply src/btree/bt_vrfy.c]: done.

File: src/btree/bt_vrfy.c (hunk 3 above)
Comment (boboeee) [reply src/btree/bt_vrfy.c]: agreed.

File: src/btree/bt_vrfy.c [hunk 7]
```diff
@@ -35,13 +35,14 @@ typedef struct {
...
 static int __verify_col_var_page_hs(WT_SESSION_IMPL *, WT_REF *, WT_VSTUFF *);
-static int __verify_key_hs(WT_SESSION_IMPL *, WT_ITEM *, WT_VSTUFF *);
+static int __verify_key_hs(WT_SESSION_IMPL *, WT_ITEM *, WT_VSTUFF *, WT_CELL_UNPACK *);
 static int __verify_page_cell(WT_SESSION_IMPL *, WT_REF *, WT_CELL_UNPACK *, WT_VSTUFF *);
 static int __verify_row_int_key_order(
   WT_SESSION_IMPL *, WT_PAGE *, WT_REF *, uint32_t, WT_VSTUFF *);
 static int __verify_row_leaf_key_order(WT_SESSION_IMPL *, WT_REF *, WT_VSTUFF *);
 static int __verify_row_leaf_page_hs(WT_SESSION_IMPL *, WT_REF *, WT_VSTUFF *);
 static int __verify_tree(WT_SESSION_IMPL *, WT_REF *, WT_CELL_UNPACK *, WT_VSTUFF *);
+static const char *__verify_timestamp_to_pretty_string(wt_timestamp_t);
```
[the remark this answers is not in the document]
Comment (boboeee) [reply src/btree/bt_vrfy.c]: agreed.

File: src/btree/bt_vrfy.c [hunk 8]
```diff
@@ -35,13 +35,14 @@ typedef struct {
 
 static void __verify_checkpoint_reset(WT_VSTUFF *);
 static int __verify_col_var_page_hs(WT_SESSION_IMPL *, WT_REF *, WT_VSTUFF *);
-static int __verify_key_hs(WT_SESSION_IMPL *, WT_ITEM *, WT_VSTUFF *);
+static int __verify_key_hs(WT_SESSION_IMPL *, WT_ITEM *, WT_VSTUFF *, WT_CELL_UNPACK *);
```
[the remark this answers is not in the document]
Comment (boboeee) [reply src/btree/bt_vrfy.c]: agreed.

File: src/btree/bt_vrfy.c [hunk 9]
```diff
@@ -232,21 +233,24 @@ __verify_row_leaf_page_hs(WT_SESSION_IMPL *session, WT_REF *ref, WT_VSTUFF *vs)
  *     Verify a key against the history store given a cell unpack holding key values.
```
[the remark this answers is not in the document]
Comment (boboeee) [reply src/btree/bt_vrfy.c]: agreed.

File: src/btree/bt_vrfy.c [hunk 10]
```diff
@@ -280,12 +284,31 @@ __verify_key_hs(WT_SESSION_IMPL *session, WT_ITEM *key, WT_VSTUFF *vs)
 #else
         WT_UNUSED(vs);
 #endif
-    }
 
+        /* Perform Verification of data continuity */
+        if (orig_start.timestamp < stop.timestamp) {
+            ts1_bp = __verify_timestamp_to_pretty_string(stop.timestamp);
```
[the remark this answers is not in the document]
Comment (boboeee) [reply src/btree/bt_vrfy.c]: agreed.

File: src/btree/bt_vrfy.c [hunk 11]
```diff
@@ -280,12 +284,31 @@ __verify_key_hs(WT_SESSION_IMPL *session, WT_ITEM *key, WT_VSTUFF *vs)
...
+              __wt_buf_set_printable(session, hs_key->data, hs_key->size, vs->tmp1), stop.txnid,
+              orig_start.txnid);
+        }
+        memcpy(&orig_start, &start, sizeof(WT_TIME_PAIR));
+    }
 err:
     __wt_scr_free(session, &hs_key);
     WT_RET(__wt_hs_cursor_close(session, &hs_cursor, session_flags));
 
+    if (ret != WT_NOTFOUND) {
```
[the remark this answers is not in the document]
Comment (boboeee) [reply src/btree/bt_vrfy.c]: agreed.

File: src/btree/bt_vrfy.c [hunk 12]
```diff
@@ -232,21 +233,24 @@ __verify_row_leaf_page_hs(WT_SESSION_IMPL *session, WT_REF *ref, WT_VSTUFF *vs)
...
 {
     WT_BTREE *btree;
     WT_CURSOR *hs_cursor;
     WT_DECL_ITEM(hs_key);
     WT_DECL_RET;
-    WT_TIME_PAIR start, stop;
+    WT_TIME_PAIR start, stop, orig_start;
     uint32_t hs_btree_id, session_flags;
     int cmp, exact;
+    const char *ts1_bp, *ts2_bp;
```
[the remark this answers is not in the document]
Comment (boboeee) [reply src/btree/bt_vrfy.c]: agreed.

File: src/btree/bt_vrfy.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (boboeee) [reply src/btree/bt_vrfy.c]: agreed.

File: src/btree/bt_vrfy.c [hunk 13]
```diff
@@ -232,21 +233,24 @@ __verify_row_leaf_page_hs(WT_SESSION_IMPL *session, WT_REF *ref, WT_VSTUFF *vs)
...
 
     btree = S2BT(session);
     hs_btree_id = btree->id;
     session_flags = 0;
     stop.timestamp = 0;
     stop.txnid = 0;
+    orig_start.timestamp = unpack->start_ts;
```
[the remark this answers is not in the document]
Comment (boboeee) [reply src/btree/bt_vrfy.c]: agreed.

File: src/btree/bt_vrfy.c [hunk 14]
```diff
@@ -280,12 +284,36 @@ __verify_key_hs(WT_SESSION_IMPL *session, WT_ITEM *key, WT_VSTUFF *vs)
...
+              stop.txnid, prev_start.txnid);
+        }
+        prev_start.timestamp = start.timestamp;
+        prev_start.txnid = start.txnid;
+    }
 err:
     __wt_scr_free(session, &hs_key);
     WT_RET(__wt_hs_cursor_close(session, &hs_cursor, session_flags));
 
+    WT_RET_NOTFOUND_OK(ret);
```
Comment (jeremytay90) [inline src/btree/bt_vrfy.c]: Sulabh encountered an issue here, so the merge conflict you have now is a fix for that.

File: src/btree/bt_vrfy.c [hunk 15]
```diff
@@ -947,6 +949,30 @@ __verify_txn_addr_cmp(WT_SESSION_IMPL *session, WT_REF *ref, uint32_t cell_num,
       gt ? "less than" : "greater than", txn2_name, txn2);
 }
 
+/*
+ * __verify_timestamp_to_pretty_string --
+ *     Convert a timestamp to a pretty string, utilises existing timestamp to string function
+ */
+static const char *
+__verify_timestamp_to_pretty_string(wt_timestamp_t ts)
```
[the remark this answers is not in the document]
Comment (jeremytay90) [reply src/btree/bt_vrfy.c]: Seems fine to me

File: src/btree/bt_vrfy.c (hunk 2 above)
Comment (jeremytay90) [reply src/btree/bt_vrfy.c]: Sounds good, maybe simplify the wording a bit so it's more consistent with other messages.

File: src/btree/bt_vrfy.c (hunk 1 above)
Comment (jeremytay90) [reply src/btree/bt_vrfy.c]: Hmm that could be annoying, but I guess there's nothing we can do about that.

**Comment by jeremytay90** [comment]
LGTM

File: src/btree/bt_vrfy.c [hunk 16]
```diff
@@ -229,21 +230,24 @@ __verify_row_leaf_page_hs(WT_SESSION_IMPL *session, WT_REF *ref, WT_VSTUFF *vs)
 
 /*
  * __verify_key_hs --
- *     Verify a key against the history store given a cell unpack holding key values.
+ *     Verify a key against the history store. Unpack will be used to verify the data continuity
```
Comment (tetsuo-cpp) [inline src/btree/bt_vrfy.c]: This comment is a bit misleading. You're only using unpack on to check one of the timestamp ranges. The remaining ones are using the time pairs embedded in the history store key to check data continuity.

File: src/btree/bt_vrfy.c [hunk 17]
```diff
@@ -229,21 +230,24 @@ __verify_row_leaf_page_hs(WT_SESSION_IMPL *session, WT_REF *ref, WT_VSTUFF *vs)
...
     WT_DECL_ITEM(hs_key);
     WT_DECL_RET;
-    WT_TIME_PAIR start, stop;
+    WT_TIME_PAIR prev_start, start, stop;
     uint32_t hs_btree_id, session_flags;
     int cmp, exact;
 
     btree = S2BT(session);
     hs_btree_id = btree->id;
+    prev_start.timestamp = unpack->start_ts;
```
Comment (tetsuo-cpp) [inline src/btree/bt_vrfy.c]: A comment would be good here to explain why we're doing this.

File: src/btree/bt_vrfy.c [hunk 18]
```diff
@@ -281,6 +285,30 @@ __verify_key_hs(WT_SESSION_IMPL *session, WT_ITEM *key, WT_VSTUFF *vs)
 #else
         WT_UNUSED(vs);
 #endif
+
+        /* Perform Verification of data continuity. */
```
Comment (tetsuo-cpp) [inline src/btree/bt_vrfy.c]: I'd prefer something like "verify that the current record's stop time pair doesn't overlap the start time pair of its successor".
